### PR TITLE
Cargo: Build CLAP command using derive

### PR DESCRIPTION
**Pull request by xFrednet**
Just a simple refactoring. At first, I was a bit hesitant, but I believe that this is the better way to handle Marker's CLI. Only the `--test-setup` argument changed into a hidden subcommand, the rest should be the same as before.

Nothing more to say otherwise. Reviews are appreciated. For everyone reading this, have a beautiful day! :D

### CODE REVIEW

File: cargo-marker/src/cli.rs
Comment (NathanFrasier) [inline cargo-marker/src/cli.rs:38]: I think you didn't finish your thought here.

File: cargo-marker/src/main.rs
Comment (NathanFrasier) [inline cargo-marker/src/main.rs:25]: I assume this was left during debugging, should be removed.

**Review by NathanFrasier — CHANGES_REQUESTED** [review]
Tested locally, looks good to me functionality wise, a few small items to polish before merge though.

**Comment by xFrednet** [comment]
Good catches, thank you very much for the review!

You can merge this PR by commenting `bors r+` if you think this version is ready. I don't want to put pressure on you. It's totally fine if you would rather not give it another review or merge it :)

bors delegate=NathanFrasier

File: cargo-marker/src/cli.rs
Comment (xFrednet) [reply cargo-marker/src/cli.rs:43]: I tried doing it with a take call at first, but that always consumed the iterator. I have the feeling, that this should be possible, but I couln't figure out how :sweat_smile:

File: cargo-marker/src/cli.rs
Comment (xFrednet) [reply cargo-marker/src/cli.rs:42]: Yeah, the other once was litteraly my brain just jumping to a different topic ^^

File: cargo-marker/src/cli.rs
Comment (NathanFrasier) [reply cargo-marker/src/cli.rs:43]: I feel like this should totally be doable, if not with the stdlib, at least with `itertools` would that be an acceptable dependency to introduce? it seems pretty low risk to me.

I'll see if I can make that happen and give a concrete code suggestion later, hopefully tonight.

Comment (xFrednet) [reply cargo-marker/src/cli.rs:43]: Adding `itertools` would be totally fine with me :+1:

Thank you for looking into this!

File: cargo-marker/src/main.rs
Comment (Veetaha) [inline cargo-marker/src/main.rs:87]: IIRC there are several places where dev build flag influences the discovery of the toolchain and driver.

I'd prefer if there was a magic env variable e.g. `MARKER_DEV_MODE=1` that could be used for conditions when a dev mode is used. This wouldn't require building in release or changing code temporarily to [de]activate the dev mode.

File: cargo-marker/src/cli.rs
Comment (Veetaha) [reply cargo-marker/src/cli.rs:43]: I guess it should be possible to express it using attributes that we want a [`trailing_var_arg`](https://docs.rs/clap/latest/clap/struct.Arg.html#method.trailing_var_arg), although I didn't try it. I also see a [cookbook for something like this](https://docs.rs/clap/latest/clap/_derive/_cookbook/escaped_positional_derive/index.html). It suggests using `#[arg(last = true)]`, and is probably use case that we have here?

Btw, there is cookbook for cargo subcommand in [clap docs](https://docs.rs/clap/latest/clap/_derive/_cookbook/cargo_example_derive/index.html), although it's not very useful.

File: cargo-marker/src/cli.rs
Comment (Veetaha) [inline cargo-marker/src/cli.rs:85]: IIRC booleans have the `false` default by default. I think you don't need to explicitly spell `default_value_t` or maybe something changed in the new version of clap? Because I am still stuck with `structopt` on my work 😓

File: cargo-marker/src/cli.rs
Comment (Veetaha) [inline cargo-marker/src/cli.rs:90]: Same thing about `#[arg(last = true)]` may apply here.

**Review by Veetaha — APPROVED** [review]

File: cargo-marker/src/main.rs
Comment (xFrednet) [reply cargo-marker/src/main.rs:87]: Part of this change was to have a unified way across the entire project to enter dev mode. Having an environment value could work, but would be a bit weird, as it would always need to be set when interacting with the project. Or am I missing another option I'm missing?

File: cargo-marker/src/cli.rs
Comment (xFrednet) [reply cargo-marker/src/cli.rs:85]: I'll try this. The PR was hacked together during travels with a bad internet connection ^^

File: cargo-marker/src/cli.rs
Comment (xFrednet) [reply cargo-marker/src/cli.rs:43]: `#[clap(last = true)]` works! :tada: 

It makes the usage message a bit weird: 

```
Usage: cargo-marker [OPTIONS] [-- <CARGO_ARGS>...] [COMMAND]
```

But that is fixed, by just keeping the `override_usage` attribute, which Marker was already using. Thank you for the suggestion :D

**Comment by xFrednet** [comment]
I want to combine the two review commits, before the merge. I'll wait a bit more for more feedback :)

File: cargo-marker/src/main.rs
Comment (Veetaha) [reply cargo-marker/src/main.rs:87]: I guess you are right. Maybe MARKER_RELEASE_MODE could then be used as an override to opt out.

**Review by Veetaha — APPROVED** [review]
I happened to stumble upon a project called `cargo-nextest`. It also uses clap derive for its CLI. That project doesn't make a special case for running `cargo-nextest` directly. It requires you to specify `cargo-nextest nextest` as a subcommand ([CLI code](https://github.com/nextest-rs/nextest/blob/1cd93240e022f28968bffbbec4e6ab7a9a585dee/cargo-nextest/src/dispatch.rs#L76)). Maybe that code could be helpful

**Comment by xFrednet** [comment]
> I happened to stumble upon a project called `cargo-nextest`. It also uses clap derive for its CLI. That project doesn't make a special case for running `cargo-nextest` directly. It requires you to specify `cargo-nextest nextest` as a subcommand ([CLI code](https://github.com/nextest-rs/nextest/blob/1cd93240e022f28968bffbbec4e6ab7a9a585dee/cargo-nextest/src/dispatch.rs#L76)). Maybe that code could be helpful

I really like this idea, thank you for the suggestion!

File: cargo-marker/src/cli.rs
Comment (xFrednet) [reply cargo-marker/src/cli.rs:101]: The `assert_maches` link points to the CLI code of next test and I can't find the create on crates.io

Something like this would definitely be useful. You already mentioned this in a previous PR and https://github.com/rust-marker/marker/issues/202 exists to investigate it. I would rather not add it in this PR, to keep it more contained, but this would be a great application for it :+1: (I just need time to get through all the issues :sweat_smile: )

Comment (Veetaha) [reply cargo-marker/src/cli.rs:111]: Yeah, an apology for the typo. It's [`assert_matches`](https://docs.rs/assert_matches/latest/assert_matches/)

Comment (xFrednet) [reply cargo-marker/src/cli.rs:111]: I think [expect_test](https://docs.rs/expect-test/latest/expect_test/) would be a better fit here, mostly because it allows automatic test updates. Still, I won't add it in this PR. My next weeks will be more busy, and I want to make sure this and the other stuff is added before 0.2.0.

File: cargo-marker/src/main.rs
Comment (xFrednet) [reply cargo-marker/src/main.rs:87]: I feel like it's too easy for a user to accidentally install it without setting the environment value.

**Comment by xFrednet** [comment]
I've squashed the review commits, this should not be ready to be merged.

bors r=Veetaha

**Pull request merged**
